// @aws-quickstart/eks-blueprints/dist/cluster-providers/asg-cluster-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsgClusterProvider = void 0;
const generic_cluster_provider_1 = require("./generic-cluster-provider");
/**
 * AsgClusterProvider provisions an EKS cluster with an autoscaling group for self-managed capacity.
 */
class AsgClusterProvider extends generic_cluster_provider_1.GenericClusterProvider {
    constructor(props) {
        var _a, _b;
        super({ ...generic_cluster_provider_1.defaultOptions, ...props, ...{
                autoscalingNodeGroups: [{
                        id: (_b = (_a = props === null || props === void 0 ? void 0 : props.id) !== null && _a !== void 0 ? _a : props === null || props === void 0 ? void 0 : props.clusterName) !== null && _b !== void 0 ? _b : "eks-blueprints-asg",
                        desiredSize: props === null || props === void 0 ? void 0 : props.desiredSize,
                        maxSize: props === null || props === void 0 ? void 0 : props.maxSize,
                        minSize: props === null || props === void 0 ? void 0 : props.minSize,
                        vpcSubnets: props === null || props === void 0 ? void 0 : props.vpcSubnets,
                        instanceType: props === null || props === void 0 ? void 0 : props.instanceType,
                        machineImageType: props === null || props === void 0 ? void 0 : props.machineImageType,
                        updatePolicy: props === null || props === void 0 ? void 0 : props.updatePolicy
                    }]
            } });
    }
}
exports.AsgClusterProvider = AsgClusterProvider;
