// @aws-quickstart/eks-blueprints/dist/cluster-providers/mng-cluster-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertEC2NodeGroup = exports.MngClusterProvider = void 0;
const generic_cluster_provider_1 = require("./generic-cluster-provider");
/**
 * MngClusterProvider provisions an EKS cluster with a managed node group for managed capacity.
 */
class MngClusterProvider extends generic_cluster_provider_1.GenericClusterProvider {
    constructor(props) {
        var _a, _b;
        super({ ...generic_cluster_provider_1.defaultOptions, ...props, ...{
                managedNodeGroups: [{
                        id: (_b = (_a = props === null || props === void 0 ? void 0 : props.id) !== null && _a !== void 0 ? _a : props === null || props === void 0 ? void 0 : props.clusterName) !== null && _b !== void 0 ? _b : "eks-blueprints-mng",
                        amiReleaseVersion: props === null || props === void 0 ? void 0 : props.amiReleaseVersion,
                        customAmi: props === null || props === void 0 ? void 0 : props.customAmi,
                        amiType: props === null || props === void 0 ? void 0 : props.amiType,
                        desiredSize: props === null || props === void 0 ? void 0 : props.desiredSize,
                        instanceTypes: props === null || props === void 0 ? void 0 : props.instanceTypes,
                        maxSize: props === null || props === void 0 ? void 0 : props.maxSize,
                        minSize: props === null || props === void 0 ? void 0 : props.minSize,
                        nodeGroupCapacityType: props === null || props === void 0 ? void 0 : props.nodeGroupCapacityType,
                        vpcSubnets: props === null || props === void 0 ? void 0 : props.vpcSubnets,
                    }]
            } });
    }
}
exports.MngClusterProvider = MngClusterProvider;
/**
 * Validates that cluster is backed by EC2 either through a managed node group or through a self-managed autoscaling group.
 * @param clusterInfo
 * @param source Used for error message to identify the source of the check
 * @returns
 */
//TODO: move to clusterInfo
function assertEC2NodeGroup(clusterInfo, source) {
    if (clusterInfo.nodeGroups != undefined && clusterInfo.nodeGroups.length > 0) {
        return clusterInfo.nodeGroups;
    }
    if (clusterInfo.autoscalingGroups != undefined && clusterInfo.autoscalingGroups.length > 0) {
        return clusterInfo.autoscalingGroups;
    }
    throw new Error(`${source} is supported with EKS EC2 only`);
}
exports.assertEC2NodeGroup = assertEC2NodeGroup;
