// @aws-quickstart/eks-blueprints/dist/utils/vpc-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tagSubnets = exports.tagSecurityGroup = void 0;
const ec2 = require("aws-cdk-lib/aws-ec2");
const custom_resources_1 = require("aws-cdk-lib/custom-resources");
/**
 * Tags EC2 Security Group with given tag and value - used for EKS Security Group Tagging
 * @param stack - CDK Stack
 * @param securityGroupId - Security Group Resource ID
 * @param key - Tag Key
 * @param value - Tag Value
 */
function tagSecurityGroup(stack, securityGroupId, key, value) {
    const tags = [{
            Key: key,
            Value: value
        }];
    const arn = `arn:${stack.partition}:ec2:${stack.region}:${stack.account}:security-group/` + securityGroupId;
    const parameters = {
        Resources: [securityGroupId],
        Tags: tags
    };
    applyEC2Tag("eks-sg", stack, parameters, key, [arn]);
}
exports.tagSecurityGroup = tagSecurityGroup;
/**
 * Tags VPC Subnets with given tag and value.
 * @param stack - CDK Stack
 * @param subnets - a list of subnets
 * @param key - Tag Key
 * @param value - Tag Value
 */
function tagSubnets(stack, subnets, key, value) {
    for (const subnet of subnets) {
        if (!ec2.Subnet.isVpcSubnet(subnet)) {
            throw new Error('This is not a valid subnet.');
        }
    }
    const tags = [{
            Key: key,
            Value: value
        }];
    const arns = subnets.map(function (val, _) {
        return `arn:${stack.partition}:ec2:${stack.region}:${stack.account}:subnet/` + val.subnetId;
    });
    const parameters = {
        Resources: subnets.map((arn) => arn.subnetId),
        Tags: tags
    };
    applyEC2Tag("subnet", stack, parameters, key, arns);
}
exports.tagSubnets = tagSubnets;
function applyEC2Tag(id, stack, parameters, tag, resources) {
    const sdkCall = {
        service: 'EC2',
        action: 'createTags',
        parameters: parameters,
        physicalResourceId: { id: `${tag}-${id}-Tagger` }
    };
    new custom_resources_1.AwsCustomResource(stack, `${id}-tags-${tag}`, {
        policy: custom_resources_1.AwsCustomResourcePolicy.fromSdkCalls({
            resources: resources,
        }),
        onCreate: sdkCall,
        onUpdate: sdkCall,
        onDelete: {
            ...sdkCall,
            action: 'deleteTags',
        },
    });
}
