// @aws-quickstart/eks-blueprints/dist/addons/secrets-store/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretsStoreAddOn = void 0;
const csi_driver_provider_aws_1 = require("./csi-driver-provider-aws");
const utils_1 = require("../../utils");
/**
 * Defaults options for the add-on
 */
const defaultProps = {
    ascpUrl: 'https://raw.githubusercontent.com/aws/secrets-store-csi-driver-provider-aws/main/deployment/aws-provider-installer.yaml',
    chart: 'secrets-store-csi-driver',
    name: 'secrets-store-csi-driver',
    namespace: 'kube-system',
    version: 'v0.0.23',
    release: 'blueprints-addon-secret-store-csi-driver',
    repository: 'https://kubernetes-sigs.github.io/secrets-store-csi-driver/charts',
    rotationPollInterval: undefined,
    syncSecrets: true,
};
class SecretsStoreAddOn {
    constructor(props) {
        this.options = (0, utils_1.cloneDeep)({ ...defaultProps, ...props });
    }
    deploy(clusterInfo) {
        const csiDriverProviderAws = new csi_driver_provider_aws_1.CsiDriverProviderAws(this.options);
        return Promise.resolve(csiDriverProviderAws.deploy(clusterInfo));
    }
}
exports.SecretsStoreAddOn = SecretsStoreAddOn;
