// @aws-quickstart/eks-blueprints/dist/addons/secrets-store/secret-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LookupSsmSecretByAttrs = exports.LookupSecretsManagerSecretByArn = exports.LookupSecretsManagerSecretByName = exports.GenerateSecretManagerProvider = void 0;
const aws_secretsmanager_1 = require("aws-cdk-lib/aws-secretsmanager");
const aws_ssm_1 = require("aws-cdk-lib/aws-ssm");
/**
 * Generate a new Secret on Secrets Manager
 */
class GenerateSecretManagerProvider {
    constructor(id, secretName) {
        this.id = id;
        this.secretName = secretName;
    }
    provide(clusterInfo) {
        const secret = new aws_secretsmanager_1.Secret(clusterInfo.cluster.stack, this.id, {
            secretName: this.secretName
        });
        return secret;
    }
}
exports.GenerateSecretManagerProvider = GenerateSecretManagerProvider;
/**
 * Lookup Secret in SecretsManager by Name
 */
class LookupSecretsManagerSecretByName {
    /**
     * @param secretName
     * @param id
     */
    constructor(secretName, id) {
        this.secretName = secretName;
        this.id = id;
    }
    provide(clusterInfo) {
        var _a;
        return aws_secretsmanager_1.Secret.fromSecretNameV2(clusterInfo.cluster.stack, (_a = this.id) !== null && _a !== void 0 ? _a : `${this.secretName}-Lookup`, this.secretName);
    }
}
exports.LookupSecretsManagerSecretByName = LookupSecretsManagerSecretByName;
/**
 * Lookup Secret in SecretsManager by Arn
 */
class LookupSecretsManagerSecretByArn {
    /**
     * @param secretArn
     * @param id
     */
    constructor(secretArn, id) {
        this.secretArn = secretArn;
        this.id = id;
    }
    provide(clusterInfo) {
        var _a;
        return aws_secretsmanager_1.Secret.fromSecretCompleteArn(clusterInfo.cluster.stack, (_a = this.id) !== null && _a !== void 0 ? _a : `${this.secretArn}-Lookup`, this.secretArn);
    }
}
exports.LookupSecretsManagerSecretByArn = LookupSecretsManagerSecretByArn;
/**
 * Lookup SSM Parameter Store Secret by Name
 */
class LookupSsmSecretByAttrs {
    /**
     * @param secretName
     * @param version
     * @param encryptionKey
     * @param simpleName
     * @param id
     */
    constructor(secretName, version, encryptionKey, simpleName, id) {
        this.secretName = secretName;
        this.version = version;
        this.encryptionKey = encryptionKey;
        this.simpleName = simpleName;
        this.id = id;
    }
    /**
     * Lookup the secret string parameter
     * @param clusterInfo
     * @returns
     */
    provide(clusterInfo) {
        var _a;
        return aws_ssm_1.StringParameter.fromSecureStringParameterAttributes(clusterInfo.cluster.stack, (_a = this.id) !== null && _a !== void 0 ? _a : `${this.secretName}-Lookup`, {
            parameterName: this.secretName,
            version: this.version,
            encryptionKey: this.encryptionKey,
            simpleName: this.simpleName
        });
    }
}
exports.LookupSsmSecretByAttrs = LookupSsmSecretByAttrs;
