// @aws-quickstart/eks-blueprints/dist/addons/kubevious/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KubeviousAddOn = void 0;
const helm_addon_1 = require("../helm-addon");
const utils_1 = require("../../utils");
/**
 * Default props to be used when creating the Helm chart
 */
const defaultProps = {
    name: "blueprints-kubevious-addon",
    namespace: "kubevious",
    chart: "kubevious",
    version: "0.9.13",
    release: "kubevious",
    repository: "https://helm.kubevious.io",
    values: {},
    ingressEnabled: false,
    kubeviousServiceType: "ClusterIP",
};
/**
 * Main class to instantiate the Helm chart
 */
class KubeviousAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    deploy(clusterInfo) {
        let values = populateValues(this.options);
        const chart = this.addHelmChart(clusterInfo, values);
        return Promise.resolve(chart);
    }
}
exports.KubeviousAddOn = KubeviousAddOn;
/**
 * populateValues populates the appropriate values used to customize the Helm chart
 * @param helmOptions User provided values to customize the chart
 */
function populateValues(helmOptions) {
    var _a;
    const values = (_a = helmOptions.values) !== null && _a !== void 0 ? _a : {};
    (0, utils_1.setPath)(values, "ingress.enabled", helmOptions.ingressEnabled);
    (0, utils_1.setPath)(values, "kubevious.service.type", helmOptions.kubeviousServiceType);
    // Generate a random password for MySQL DB root user
    (0, utils_1.setPath)(values, "mysql.generate_passwords", true);
    return values;
}
