// @aws-quickstart/eks-blueprints/dist/resource-providers/certificate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateCertificateProvider = exports.ImportCertificateProvider = void 0;
const acm = require("aws-cdk-lib/aws-certificatemanager");
/**
 * Certificate provider that imports certificate into the current stack by arn.
 */
class ImportCertificateProvider {
    constructor(certificateArn, id) {
        this.certificateArn = certificateArn;
        this.id = id;
    }
    provide(context) {
        return acm.Certificate.fromCertificateArn(context.scope, this.id, this.certificateArn);
    }
}
exports.ImportCertificateProvider = ImportCertificateProvider;
/**
 * Certificate provider that creates a new certificate.
 * Expects a hosted zone to be registed for validation.
 */
class CreateCertificateProvider {
    /**
     * Creates the certificate provider.
     * @param name Name of this resource that other resource providers, add-ons and teams can use for look-up.
     * @param domainName
     * @param hostedZoneResourceName
     */
    constructor(name, domainName, hostedZoneResourceName) {
        this.name = name;
        this.domainName = domainName;
        this.hostedZoneResourceName = hostedZoneResourceName;
    }
    provide(context) {
        const hostedZone = context.get(this.hostedZoneResourceName);
        return new acm.Certificate(context.scope, this.name, {
            domainName: this.domainName,
            validation: acm.CertificateValidation.fromDns(hostedZone),
        });
    }
}
exports.CreateCertificateProvider = CreateCertificateProvider;
