// @aws-quickstart/eks-blueprints/dist/resource-providers/hosted-zone.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelegatingHostedZoneProvider = exports.ImportHostedZoneProvider = exports.LookupHostedZoneProvider = void 0;
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const r53 = require("aws-cdk-lib/aws-route53");
/**
 * Simple lookup host zone provider
 */
class LookupHostedZoneProvider {
    /**
     * @param hostedZoneName name of the host zone to lookup
     * @param id  optional id for the structure (for tracking). set to hostzonename by default
     */
    constructor(hostedZoneName, id) {
        this.hostedZoneName = hostedZoneName;
        this.id = id;
    }
    provide(context) {
        var _a;
        return r53.HostedZone.fromLookup(context.scope, (_a = this.id) !== null && _a !== void 0 ? _a : `${this.hostedZoneName}-Lookup`, { domainName: this.hostedZoneName });
    }
}
exports.LookupHostedZoneProvider = LookupHostedZoneProvider;
/**
 * Direct import hosted zone provider, based on a known hosted zone ID.
 * Recommended method if hosted zone id is known, as it avoids extra look-ups.
 */
class ImportHostedZoneProvider {
    constructor(hostedZoneId, id) {
        this.hostedZoneId = hostedZoneId;
        this.id = id;
    }
    provide(context) {
        var _a;
        return r53.HostedZone.fromHostedZoneId(context.scope, (_a = this.id) !== null && _a !== void 0 ? _a : `${this.hostedZoneId}-Import`, this.hostedZoneId);
    }
}
exports.ImportHostedZoneProvider = ImportHostedZoneProvider;
/**
 * Delegating provider is a convenience approach to have a global hosted zone record in a centralized
 * account and subdomain records in respective workload accounts.
 *
 * The delegation part allows routing subdomain entries to the child hosted zone in the workload account.
 */
class DelegatingHostedZoneProvider {
    constructor(options) {
        this.options = options;
    }
    provide(context) {
        const stack = context.scope;
        const subZone = new r53.PublicHostedZone(stack, `${this.options.subdomain}-SubZone`, {
            zoneName: this.options.subdomain
        });
        if (this.options.wildcardSubdomain) {
            new r53.CnameRecord(stack, `${this.options.subdomain}-cname`, {
                zone: subZone,
                domainName: `${this.options.subdomain}`,
                recordName: `*.${this.options.subdomain}`
            });
        }
        // 
        // import the delegation role by constructing the roleArn.
        // Assuming the parent account has the delegating role with 
        // trust relationship setup to the child account.
        //
        const delegationRoleArn = stack.formatArn({
            region: '',
            service: 'iam',
            account: this.options.parentDnsAccountId,
            resource: 'role',
            resourceName: this.options.delegatingRoleName
        });
        const delegationRole = aws_iam_1.Role.fromRoleArn(stack, 'DelegationRole', delegationRoleArn);
        // create the record
        new r53.CrossAccountZoneDelegationRecord(stack, `${this.options.subdomain}-delegate`, {
            delegatedZone: subZone,
            parentHostedZoneName: this.options.parentDomain,
            delegationRole
        });
        return subZone;
    }
}
exports.DelegatingHostedZoneProvider = DelegatingHostedZoneProvider;
