// @aws-quickstart/eks-blueprints/dist/addons/argocd/manifest-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createUserNameSecretRef = exports.createSshSecretRef = exports.createSecretRef = void 0;
const __1 = require("..");
/**
 * Creates CsiSecretProps that contains secret template for ssh/username/pwd credentials.
 * In each case, the secret is expected to be a JSON structure containing url and either sshPrivateKey
 * or username and password attributes.
 * @param credentialsType SSH | USERNAME | TOKEN
 * @param secretName
 * @returns
 */
function createSecretRef(credentialsType, secretName) {
    switch (credentialsType) {
        case "SSH":
            return createSshSecretRef(secretName);
        case "USERNAME":
        case "TOKEN":
            return createUserNameSecretRef(secretName);
        default:
            throw new Error(`credentials type ${credentialsType} is not supported by ArgoCD add-on.`);
    }
}
exports.createSecretRef = createSecretRef;
/**
 * Local function to create a secret reference for SSH key.
 * @param url
 * @param secretName
 * @returns
 */
function createSshSecretRef(secretName) {
    return {
        secretProvider: new __1.LookupSecretsManagerSecretByName(secretName),
        jmesPath: [{ path: "url", objectAlias: "url" }, { path: "sshPrivateKey", objectAlias: "sshPrivateKey" }],
        kubernetesSecret: {
            secretName: secretName,
            labels: { "argocd.argoproj.io/secret-type": "repo-creds" },
            data: [
                { key: "url", objectName: "url" },
                { key: "sshPrivateKey", objectName: "sshPrivateKey" }
            ]
        }
    };
}
exports.createSshSecretRef = createSshSecretRef;
/**
 * Local function to a secret reference for username/pwd or username/token key.
 * @param url
 * @param secretName
 * @returns
 */
function createUserNameSecretRef(secretName) {
    return {
        secretProvider: new __1.LookupSecretsManagerSecretByName(secretName),
        jmesPath: [{ path: "url", objectAlias: "url" }, { path: "username", objectAlias: "username" }, { path: "password", objectAlias: "password" }],
        kubernetesSecret: {
            secretName: secretName,
            labels: new Map([
                ["argocd.argoproj.io/secret-type", "repo-creds"]
            ]),
            data: [
                { key: "url", objectName: "url" },
                { key: "username", objectName: "username" },
                { key: "password", objectName: "password" }
            ]
        }
    };
}
exports.createUserNameSecretRef = createUserNameSecretRef;
